src/fsharp/ParseAndCheckInputs.fs

```diff
@@ -393,31 +393,95 @@ let ParseOneInputLexbuf (tcConfig: TcConfig, lexResourceManager, conditionalComp
+    let lexbuf = UnicodeLexing.StreamReaderAsLexbuf(tcConfig.langVersion.SupportsFeature, reader)
 
-            // Parse the file drawing tokens from the lexbuf
-            ParseOneInputLexbuf(tcConfig, lexResourceManager, conditionalCompilationDefines, lexbuf, filename, isLastCompiland, errorLogger)
-       else
-           error(Error(FSComp.SR.buildInvalidSourceFileExtension(SanitizeFileName filename tcConfig.implicitIncludeDir), rangeStartup))
+    // Parse the file drawing tokens from the lexbuf
+    ParseOneInputLexbuf(tcConfig, lexResourceManager, conditionalCompilationDefines, lexbuf, filename, isLastCompiland, errorLogger)
 
+/// Parse an input from disk
+let ParseOneInputFile (tcConfig: TcConfig, lexResourceManager, conditionalCompilationDefines, filename, isLastCompiland, errorLogger, retryLocked) =
+    try 
+       checkInputFile tcConfig filename
+       parseInputFileAux(tcConfig, lexResourceManager, conditionalCompilationDefines, filename, isLastCompiland, errorLogger, retryLocked)
     with e -> 
         errorRecovery e rangeStartup
         EmptyParsedInput(filename, isLastCompiland) 
 
+/// Parse multiple input files from disk
+let ParseInputFiles (tcConfig: TcConfig, lexResourceManager, conditionalCompilationDefines, sourceFiles, errorLogger: ErrorLogger, exiter: Exiter, createErrorLogger: (Exiter -> CapturingErrorLogger), retryLocked) =
+    try
+        let isLastCompiland, isExe = sourceFiles |> tcConfig.ComputeCanContainEntryPoint
+        let sourceFiles = isLastCompiland |> List.zip sourceFiles |> Array.ofSeq
+
+        if tcConfig.concurrentBuild then
+            let mutable exitCode = 0
+            let delayedExiter = 
+                { new Exiter with
+                        member this.Exit n = exitCode <- n; raise StopProcessing }
+
+            // Check input files and create delayed error loggers before we try to parallel parse.
+            let delayedErrorLoggers =
+                sourceFiles
+                |> Array.map (fun (filename, _) ->
+                    checkInputFile tcConfig filename
+                    createErrorLogger(delayedExiter)
+                )
+
+            let commitDelayedErrorLoggers () =
+                delayedErrorLoggers
+                |> Array.iter (fun delayedErrorLogger ->
+                    delayedErrorLogger.CommitDelayedDiagnostics errorLogger
+                )
+
+            let results =
+                try
+                    sourceFiles 
+                    |> ArrayParallel.mapi (fun i (filename, isLastCompiland) ->
+                        let delayedErrorLogger = delayedErrorLoggers.[i]
+                
+                        let directoryName = Path.GetDirectoryName filename
+                        let input = parseInputFileAux(tcConfig, lexResourceManager, conditionalCompilationDefines, filename, (isLastCompiland, isExe), delayedErrorLogger, retryLocked)
+                        (input, directoryName)
+                    )
+                with
```

Comment:
This nested try/with with the explicit catch of StopProcessing exception and three separate calls to commitDelayedErrorLoggers feels odd, a code smell, though it's not too bad I guess. I think however there are existing patterns to cover these cases of catching StopProcessing exception.

And what exception does ArrayParallel.mapi even raise if multiple computations raise an exception?  Isn't it AggregateException or something?

Reply:
Yeah we don't automatically flatten aggregate exceptions I think

Reply:
Ok, I've flattened the aggregate exceptions if the count is one now in ArrayParallel.

Also, I used a try/finally to commit the delayed diagnostic results, so now it's just one place.